# PROBLEM

Given 'Classification.csv', build a suitable model to predict loan defaults.

# SOLUTION

split data [train:test::70:30] <br/>
build model <br/>
find out accuracy

In [1]:
import os
import pandas as pd
from sklearn.model_selection import train_test_split
import statsmodels.formula.api as sm

In [2]:
os.getcwd()

'C:\\Users\\praveen\\DS_Notebooks\\hgwxx7'

In [3]:
gc = pd.read_csv('Classification.csv')

In [4]:
gc.head()

Default checkingstatus1  duration history purpose  amount savings employ  \
0        0             A11         6     A34     A43    1169     A65    A75   
1        1             A12        48     A32     A43    5951     A61    A73   
2        0             A14        12     A34     A46    2096     A61    A74   
3        0             A11        42     A32     A42    7882     A61    A74   
4        1             A11        24     A33     A40    4870     A61    A73   

   installment status   ...   residence  property age  otherplans housing  \
0            4    A93   ...           4      A121  67        A143    A152   
1            2    A92   ...           2      A121  22        A143    A152   
2            2    A93   ...           3      A121  49        A143    A152   
3            2    A93   ...           4      A122  45        A143    A153   
4            3    A93   ...           4      A124  53        A143    A153   

  cards   job liable  tele foreign  
0     2  A173      1  A192    A201  
1     1  A173      1  A191    A201  
2     1  A172      2  A191    A201  
3     1  A173      2  A191    A201  
4     2  A173      2  A191    A201  

[5 rows x 21 columns]

# Logistic Regression

In [5]:
# random_state=123 - acts a seed for multiple iterations
# to have same data for train and test
train, test = train_test_split(gc, test_size=0.30, random_state=123)

In [6]:
model_log = sm.glm(formula='Default~duration', data=train)
model_fit_log = model_log.fit()

In [7]:
model_fit_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                Default   No. Observations:                  700
Model:                            GLM   Df Residuals:                      698
Model Family:                Gaussian   Df Model:                            1
Link Function:               identity   Scale:               0.194192112852958
Method:                          IRLS   Log-Likelihood:                -418.64
Date:                Fri, 01 Jun 2018   Deviance:                       135.55
Time:                        01:24:24   Pearson chi2:                     136.
No. Iterations:                     2                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.1040      0.034      3.062      0.002       0.037       0.171
duration       0.0089      0.001      6.136      0.000       0.006       0.012
==============================================================================
"""

# Prediction

In [8]:
def_pred = model_fit_log.predict(test["duration"])

In [9]:
def_pred

131    0.426019
203    0.211364
50     0.318692
585    0.265028
138    0.238196
632    0.238196
328    0.426019
461    0.238196
924    0.318692
479    0.238196
195    0.184533
221    0.211364
209    0.211364
378    0.426019
156    0.184533
327    0.318692
761    0.265028
320    0.372356
274    0.372356
417    0.265028
372    0.193477
299    0.291860
229    0.318692
896    0.291860
594    0.318692
983    0.426019
736    0.318692
85     0.211364
882    0.372356
7      0.426019
         ...   
107    0.211364
522    0.533347
799    0.184533
739    0.372356
348    0.157701
396    0.211364
709    0.184533
294    0.533347
995    0.211364
968    0.479683
671    0.426019
351    0.184533
958    0.354468
367    0.265028
421    0.211364
752    0.211364
237    0.291860
809    0.238196
778    0.426019
903    0.238196
914    0.318692
887    0.533347
946    0.318692
723    0.184533
240    0.318692
556    0.265028
598    0.265028
458    0.157701
586    0.184533
178    0.211364
Length: 300, dtype: floa

In [10]:
# function to segregate probabilities 

def isPro(i):
    # 0.5 depends on requirement
    if (i>=0.5):
        return 1
    else:
        return 0

In [11]:
test["def_pred_class"] = def_pred.apply(isPro)

E:\Python\Anaconda\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [12]:
test["def_pred_class"]

131    0
203    0
50     0
585    0
138    0
632    0
328    0
461    0
924    0
479    0
195    0
221    0
209    0
378    0
156    0
327    0
761    0
320    0
274    0
417    0
372    0
299    0
229    0
896    0
594    0
983    0
736    0
85     0
882    0
7      0
      ..
107    0
522    1
799    0
739    0
348    0
396    0
709    0
294    1
995    0
968    0
671    0
351    0
958    0
367    0
421    0
752    0
237    0
809    0
778    0
903    0
914    0
887    1
946    0
723    0
240    0
556    0
598    0
458    0
586    0
178    0
Name: def_pred_class, Length: 300, dtype: int64

In [13]:
pd.crosstab(test["def_pred_class"],test["Default"])

Default           0   1
def_pred_class         
0               187  85
1                13  15

In [14]:
# Accuracy = correct predictions/total
# diagonal elements are accurate predictions
# correct predictions = 187+15 = 202
# total = 187+85+13+15 = 300

In [15]:
accuracy = round(((187+15)/(187+85+13+15))*100,2)

In [16]:
accuracy

67.33

# CONCLUSION

Model has an accuracy of 67.33% to predict loan defaulters

# Points to consider

1. Cross validation - to find ideal accuracy!